### ISSUE TITLE
Istio's 1.8 multi-cluster multi-primary multi-network not working properly

### ISSUE THREAD

**Issue by israel-hdez** [istio/istio#28381]
**Bug description**

After configuring two clusters, each one in it's own pod network and own control plane, traffic is not properly load-balanced between the two clusters.

_My guess:_

When verifying installation (as in https://preliminary.istio.io/latest/docs/setup/install/multicluster/verify/), I can see that traffic is flowing OK within the same cluster, but some requests fail with `upstream connect error`.

Apparently endpoints are correctly detected for local & remote endpoints. Load balancing is, apparently, happening OK. For failed requests, I can see that sidecar logs show the IP address of remote pods. So, **I think** the sidecar is trying to push traffic as in a flat-network topology instead of routing traffic through the eastwest ingress gateway.

[x] Docs
[x] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

Istio should use eastweast ingress gateways for cross-cluster communications.

**Steps to reproduce the bug**

Used instructions in https://preliminary.istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/.

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**

```
$ istioctl version --remote
client version: 1.8.0-alpha.0
control plane version: 1.8.0-alpha.0
data plane version: 1.8.0-alpha.0 (4 proxies)

$ kubectl version --short
Client Version: v1.18.2
Server Version: v1.18.10
```

**How was Istio installed?**

Used instructions in https://preliminary.istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/.

**Environment where bug was observed (cloud vendor, OS, etc)**

Bare metal.

**Comment by stevenctl** [comment]
Could you dump the endpoints on a source pod from each cluster by hitting `127.0.0.1:15000/clusters` or using `istioctl proxy-config endpoint <pod>.<namespace>`?

**Comment by israel-hdez** [comment]
@stevenctl Yes... using ` 127.0.0.1:15000/clusters` very large file!

Attaching: [endpoints.txt](https://github.com/istio/istio/files/5454904/endpoints.txt)

**Comment by stevenctl** [comment]
Just ran through the steps again myself on GKE and could not reproduce this. 

Can you confirm the following outputs each cluster's network? 

```
for  CTX in "$CTX_CLUSTER1" "$CTX_CLUSTER2"; do 
    kubectl --context="$CTX" -n istio-system get svc istio-eastwestgateway -ojsonpath='{.metadata.labels.topology\.istio\.io/network}'
done 
```

My concerns is that the [step to deploy the eastwest gateway](https://preliminary.istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/#install-the-east-west-gateway-in-cluster1) may have some issue or another and continue. One example is if the command isn't run in the wrong directory (samples isn't available). We'll continue to do `istioctl generate` without the eastwest yaml.

**Comment by israel-hdez** [comment]
Hi, @stevenctl 

Not, using contexts, but different KUBECONFIG files. But yes, the command on each cluster is outputting the right networks:

```
% kubectl get nodes && kubectl -n istio-system get svc istio-eastwestgateway -ojsonpath='{.metadata.labels.topology\.istio\.io/network}'
NAME       STATUS   ROLES    AGE    VERSION
kukulcan   Ready    master   2d9h   v1.18.3
network1%
```

then the other node:

```
% kubectl get nodes && kubectl -n istio-system get svc istio-eastwestgateway -ojsonpath='{.metadata.labels.topology\.istio\.io/network}'
NAME    STATUS   ROLES    AGE    VERSION
tzotz   Ready    master   2d3h   v1.18.3
network2%
```

---

I'm curious... and also giving some more context...

In that same section you pointed (the [step to deploy the eastwest gateway](https://preliminary.istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/#install-the-east-west-gateway-in-cluster1)), there is the mention to wait for the IP address of the east-west gw. I have this:

```
kubectl get nodes && kubectl -n istio-system get svc istio-eastwestgateway                                            
NAME       STATUS   ROLES    AGE    VERSION
kukulcan   Ready    master   2d9h   v1.18.3
NAME                    TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)                                                           AGE
istio-eastwestgateway   LoadBalancer   10.105.93.235   192.168.110.12   15021:31911/TCP,15443:31238/TCP,15012:30960/TCP,15017:32645/TCP   7h17m

...and the other node...
kubectl get nodes && kubectl -n istio-system get svc istio-eastwestgateway                                            
NAME    STATUS   ROLES    AGE     VERSION
tzotz   Ready    master   2d16h   v1.18.3
NAME                    TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)                                                           AGE
istio-eastwestgateway   LoadBalancer   10.108.220.74   192.168.110.22   15021:32030/TCP,15443:31932/TCP,15012:32376/TCP,15017:31232/TCP   20h
```

You can see that in both cases, there is an external IP and it's reachable between clusters.
However, in other setup scenarios, the docs use the loadbalancer IP to get the public endpoint. Like in the [remote-primary case](https://preliminary.istio.io/latest/docs/setup/install/multicluster/primary-remote/#configure-cluster2-as-a-remote), you give this command:

```
export DISCOVERY_ADDRESS=$(kubectl \
    --context="${CTX_CLUSTER1}" \
    -n istio-system get svc istio-eastwestgateway \
    -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```

I'm wondering if Istio _strictly needs the IP to be allocated by a LoadBalancer_ for this to work??? I'm asking because the output of this previous command is blank in my setup and is expected because I'm assigning IPs using [the externalIPs stanza of the kubernetes Service](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#servicespec-v1-core). My setup is bare-metal and I don't have LoadBalancers. So, if gateway external IP detection is based on LoadBalancers, I don't think MC will will work property for the multi-network scenario. Perhaps, I'll need to somehow manually feed Istio with the IPs of the gateways and not rely on IP auto-discovery.

FWIW, the following command gives the externalIP on my setup:

```
kubectl get nodes && kubectl -n istio-system get svc istio-eastwestgateway -ojsonpath='{.spec.externalIPs[0]}'
```

**Comment by stevenctl** [comment]
Ah, that would explain it. For the time being inn your use-case you'll need to configure `meshNetworks`.

You can either edit the `istio` config map in `istio-system` or reinstall putting the following yaml under `global` in your installation yaml. 

```
meshNetworks:
       network1:
         endpoints:
           - fromRegistry: cluster1
         gateways:
           - address: <externalip>
             port: 15443
       network2:
         endpoints:
           - fromRegistry: cluster2
         gateways:
           - address: <externalip>
             port: 15443
````

 I think we can add support for auto-detecting externalIP although I don't know if it will make the initial 1.8 release.

**Comment by israel-hdez** [comment]
@stevenctl OK, let me try.
Do I have to specify both networks on both clusters?

**Comment by stevenctl** [comment]
Yes, this config needs to be identical in all primary clusters.

**Comment by stevenctl** [comment]
Support for externalIPs will be included in 1.8

**Comment by israel-hdez** [comment]
> Support for externalIPs will be included in 1.8

That's nice :)
And FWIW, manually configuring `meshNetworks` works fine!

**Comment by israel-hdez** [comment]
I'm closing, as I think there is no work left.
I'll try without the `meshNetworks` configuration once the final 1.8 is released (or a new alpha is released, whatever comes first)

Thanks for your help!
